### PR TITLE
refactor: executor interface and storage dependency

### PR DESCRIPTION
**Pull request by frrist**
### What
The goal of this PR is to simplify the implementation #2637 by removing the requirement of configuring storage providers for executors. It includes several refactors to the bid strategies, executors interface, and corresponding implementation.
- Executors no longer contain storage providers. Instead the compute node responsible for orchestrating executors maintains the set of available storage providers. The compute node makes storage available to executors as local volumes. 
- ParallelPrepareStorage now returns a slice of prepared storage (instead of a map). This allows the returned `PreparedStorage` type to be serialized and passes across RPC boundaries.
- `NewModuleLoader` will not fetch remote wasm modules, and instead operates over a slice of `PreparedStorage`.
  - This change was required to remove the Storage Provider type from the wasm executor which previously fetched remote content from `Job.Input` during execution.
- Modifies the `Executor` interface by removing methods related to providing storage and reducing the complexity of parameters passes to the `Run` method.

### CODE REVIEW

File: pkg/compute/executor.go
Comment (frrist) [inline pkg/compute/executor.go:102]: I hate special casing the wasm engine, but for the time being this is workable since wasm is a "builtin" executor. This change fetches wasm modules on the compute node (from either IPFS or a URL) and makes them available locally before running the wasm executor. Previously the wasm executor was making network calls to compile the wasm binary before executing it in process.

File: pkg/executor/docker/executor.go
Comment (frrist) [inline pkg/executor/docker/executor.go:47]: Docker executor no longer contains a storage provider and thus no longer needs to (or can) implement the `HasStorageLocally` and `GetVolumeSize` methods.

File: pkg/executor/docker/executor.go
Comment (frrist) [inline pkg/executor/docker/executor.go:192]: Need to address this before merging, I don't think it's necessary and I can't find any usage of the `BACALHAU_JOB_SPEC` environment variable. I believe there may have been some docker images that required access to the full job spec for some reason - I have linked to the commit that introduced this change in the code comment for the reviewer.

File: pkg/storage/parallel.go
Comment (frrist) [inline pkg/storage/parallel.go:25]: Returns a slice now as the returned map keyed with pointers cannot cross an RPC boundary.

File: pkg/executor/wasm/executor.go
Comment (rossjones) [inline pkg/executor/wasm/executor.go:89]: I think @simonwo is working on the dir/file breakdown, so just adding this comment to let him know when he's back from vacation.

File: pkg/executor/wasm/executor.go
Comment (rossjones) [inline pkg/executor/wasm/executor.go:208]: Not a big deal, but I think I preferred the old way because this is quite dense in the middle of this builder.

File: pkg/executor/wasm/loader.go
Comment (rossjones) [inline pkg/executor/wasm/loader.go:74]: We should probably correct the method name and method command if we're only returning a single module (previously this method returned several).

File: pkg/storage/parallel.go
Comment (rossjones) [inline pkg/storage/parallel.go:31]: (nit) Instead of shadowing the same name (I first read it as 1 and got confused), should we give it a different name and capture that in the function instead?

**Review by rossjones — APPROVED** [review]
Couple of comments that may/may-not be useful, and one opportunity to move a bit more code out of the executor (creating output dir) but LGTM, definitely a positive change.

File: pkg/executor/docker/executor.go
Comment (frrist) [reply pkg/executor/docker/executor.go:145]: Agreed! Would it be alright if land this as a follow on? 

I think we'll need to do some (re)thinking regarding a potentially different `Output` type, but am unsure what it should resemble. As an example, the wasm executor requires an `Output` to have a `Path`, but doesn't use it - it just joins the results directory with the output name. The docker executor does similar, but uses the `Path` as a target for it's mounted filesystem. 

I think a more ideal case would be passing to each executor a set of inputs and set of outputs as interfaces to a filesystem, but we also need a way to serialize this information across the RPC used by a compute node to orchestrate and executor.

File: pkg/executor/types.go
Comment (frrist) [reply pkg/executor/types.go:35]: Would that mean EnvironmentVariables becoming its own field on the JobSpec? Generally seems fine to me.

File: pkg/executor/wasm/loader.go
Comment (frrist) [reply pkg/executor/wasm/loader.go:74]: Good catch, will also make it private as it's only used internally.

File: pkg/executor/types.go
Comment (wdbaruni) [reply pkg/executor/types.go:35]: I also like the idea of moving Envs outside of the engine spec to the `Job.Spec` level. My only concern would be if an executor and a publisher within the same spec want to use different values for the same env variable. We don't support Envs in publishers today, so it may not be an immediate concern

File: pkg/executor/docker/network.go
Comment (wdbaruni) [inline pkg/executor/docker/network.go:99]: nit: helpful to rename this to jobID

**Review by wdbaruni — APPROVED** [review]
Just minor comments that shouldn't block the PR

**Pull request merged**
